### DISCUSSION TITLE
Using Reqnroll with Playwright: any drawbacks or limitations?

### DISCUSSION THREAD

**Discussion by thomasvdr** [reqnroll/Reqnroll#209] [Q&A] [answered]
Hello,

We are considering the usage of Reqnroll with Playwright.

We've seen the sample of ZsoltDunai and we have no problem to run it.
However, we are wondering if there are any drawbacks/limitations of using Reqnroll with Playwright. 
Like, things that don't pop up when using a quick sample but could be a limitation when writing more complexes scenarios or using the differents features of playwright (autowait, headless/headful, parallel execution,...).
Are you aware of such limitations that we should consider?

Additional question: in the sample of ZsoltDunai, we see that they create a new Browser instance for each Scenario while, [in a discussion in the Specflow repo](https://github.com/SpecFlowOSS/SpecFlow/issues/706#issuecomment-960553909), Timm Krauss creates one instance of the browser for each test run and then a new page for each Scenario (I think for the sake of parallel execution). 
Would you recommend to do the same with Reqnroll?

Thank you for your help.

**Comment by gasparnagy** [comment, marked as answer]
It makes sense to specify what do we mean on "Playwright" here, because it is actually two things:

- a generic browser automation library ported to many platforms (like .NET)
- a node.js based test runner with a lot of fancy features, like parallel handling and reporting

The Playwright test runner can only be used with node.js and JavaScript/TypeScript, so if you choose to use the Playwright .NET library, you automatically loose those benefits.

Once you have decided to use the Playwright .NET library, driving that with Reqnroll does not cause any drawbacks that you would need to accept. As far as I know all functionalities of the library can be used from Reqnroll as well.

The browser object pooling (how/when do you reuse the browser) is definitely something you should consider and find a solution for. For that, many here suggested to use some sort of pooling using the `Microsoft.Extensions.ObjectPool` package (see also https://github.com/orgs/reqnroll/discussions/210), so this is what I would recommend as well.

In the [ReqOverflow sample](https://github.com/reqnroll/Sample-ReqOverflow/tree/main/ReqOverflow.Web) I was using another pooling approach that uses the built-in infrastructure of Reqnroll. This is also something you can consider, but with the `Microsoft.Extensions.ObjectPool` package you get more configuration options and reduce the "magic factor" (because the Reqnroll infra is not very documented).
